sympy/solvers/ode/systems.py

```diff
@@ -418,6 +439,48 @@ def _linear_neq_order1_type1(match_):
     return sol_dict
 
 
+def _linear_neq_order1_type2(match_):
+    r"""
+    System of n first-order coefficient linear non-homogeneous differential equations
+
+    .. math::
+        X' = A X + b(t)
+
+    where $X$ is the vector of $n$ dependent variables, $t$ is the dependent variable, $X'$
+    is the first order differential of $X$ with respect to $t$, $A$ is a $n \times n$
+    constant coefficient matrix and $b(t)$ is the non-homogeneous term.
+
+    The solution of the above system is:
+
+    .. math::
+        X = e^{A t} ( \int e^{- A t} b \,dt + C)
+
+    where $C$ is the vector of constants.
+
+    """
+    eq = match_['eq']
+    func = match_['func']
+    fc = match_['func_coeff']
+    b = match_['rhs']
+
+    n = len(eq)
+    t = list(list(eq[0].atoms(Derivative))[0].atoms(Symbol))[0]
+    constants = numbered_symbols(prefix='C', cls=Symbol, start=1)
+
+    # This needs to be modified in future so that fc is only of type Matrix
+    M = -fc if type(fc) is Matrix else Matrix(n, n, lambda i,j:-fc[i,func[j],0])
+
+    P, J = matrix_exp_jordan_form(M, t)
+    P = simplify(P)
+    Cvect = Matrix(list(next(constants) for _ in range(n)))
+    sol_vector = P * J * (integrate(J.inv() * P.inv() * b, t) + Cvect)
+
+    sol_vector = [collect(expand_mul(s), ordered(J.atoms(exp)), exact=True) for s in sol_vector]
```

Comment:
> Is it the call to integrate that causes the exponentials to become expanded?

This needs to be checked but it's likely that it is the case.

Also, I tried using `powsimp` function but this function causes the exponential to be expanded and before using this function, the exponentials are fine for one of the test cases.

Reply:
Using the generators from the solutions did cause the solutions to be shorter than before.

Reply:
When does `powsimp` expand the exponentials? Can you give an example?

The purpose of `powsimp` is to combine powers not expand them.

Reply:
@oscarbenjamin The `powsimp` wasn't the reason for expanded exponentials, it was because of generators, but now they are working properly and `powsimp` does help combine the powers.